packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js

```diff
@@ -0,0 +1,463 @@
+
+    if (isLastBlockImageOrCode(editor)) {
+      // insert blank line to add new blocks below code or image blocks
+      Transforms.insertNodes(
+        editor,
+        {
+          type: 'paragraph',
+          children: [{ type: 'text', text: '' }],
+        },
+        { at: [editor.children.length] }
+      );
+    }
+
+    if (optionKey === 'image') {
+      setIsMediaLibraryVisible(true);
+    }
+  };
+
+  return (
+    <>
+      <Select
+        startIcon={<Icon as={blocks[blockSelected].icon} />}
+        onChange={selectOption}
+        placeholder={blocks[blockSelected].label}
+        value={blockSelected}
+        aria-label={formatMessage({
+          id: 'components.Blocks.blocks.selectBlock',
+          defaultMessage: 'Select a block',
+        })}
+      >
+        {blockKeysToInclude.map((key) => (
+          <BlockOption
+            key={key}
+            value={key}
+            label={blocks[key].label}
+            icon={blocks[key].icon}
+            matchNode={blocks[key].matchNode}
+            handleSelection={setBlockSelected}
+            blockSelected={blockSelected}
+          />
+        ))}
+      </Select>
+      {isMediaLibraryVisible && <ImageDialog handleClose={() => setIsMediaLibraryVisible(false)} />}
+    </>
+  );
+};
+
+const BlockOption = ({ value, icon, label, handleSelection, blockSelected, matchNode }) => {
+  const { formatMessage } = useIntl();
+  const editor = useSlate();
+
+  const isActive = isBlockActive(editor, matchNode);
+  const isSelected = value === blockSelected;
+
+  React.useEffect(() => {
+    if (isActive && !isSelected) {
+      handleSelection(value);
+    }
+  }, [handleSelection, isActive, isSelected, value]);
+
+  return (
+    <Option
+      startIcon={<Icon as={icon} color={isSelected ? 'primary600' : 'neutral600'} />}
+      value={value}
+    >
+      {formatMessage(label)}
+    </Option>
+  );
+};
+
+BlockOption.propTypes = {
+  icon: PropTypes.elementType.isRequired,
+  value: PropTypes.string.isRequired,
+  label: PropTypes.shape({
+    id: PropTypes.string.isRequired,
+    defaultMessage: PropTypes.string.isRequired,
+  }).isRequired,
+  matchNode: PropTypes.func.isRequired,
+  handleSelection: PropTypes.func.isRequired,
+  blockSelected: PropTypes.string.isRequired,
+};
+
+const ListButton = ({ icon, format, label }) => {
+  const editor = useSlate();
+
+  /**
+   *
+   * @param {import('slate').Node} node
+   * @returns boolean
+   */
+  const isListNode = (node) => {
+    return !Editor.isEditor(node) && SlateElement.isElement(node) && node.type === 'list';
+  };
+
+  const isListActive = () => {
+    const { selection } = editor;
+
+    if (!selection) return false;
+
+    const [match] = Array.from(
+      Editor.nodes(editor, {
+        at: Editor.unhangRange(editor, selection),
+        match: (node) => isListNode(node) && node.format === format,
+      })
+    );
+
+    return Boolean(match);
+  };
+
+  const isActive = isListActive();
```

Comment:
This feels weird, how come we're doing it this way?

Reply:
We followed the Slate documentation in this case https://github.com/ianstormtaylor/slate/blob/main/site/examples/richtext.tsx you can find the isBlockIsActive which is similar
```
const isBlockActive = (editor, format, blockType = 'type') => {
  const { selection } = editor
  if (!selection) return false

  const [match] = Array.from(
    Editor.nodes(editor, {
      at: Editor.unhangRange(editor, selection),
      match: n =>
        !Editor.isEditor(n) &&
        SlateElement.isElement(n) &&
        n[blockType] === format,
    })
  )

  return !!match
}
```